keps/sig-api-machinery/5366-graceful-leader-transition/README.md

```diff
@@ -0,0 +1,940 @@
+shutdown controllers based on cluster state (such as the registration of
+resources that declare how a CRD should be reconciled). The extension developer
+requires that controllers shutdown gracefully so that only the controller loops
+that SHOULD be running continue to run, and that no resources are leaked over
+time as controllers are started and stopped.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Kubernetes leader elected controllers and the scheduler have been running
+without graceful shutdown for years. 
+
+Risk 1: Resource exhaustion: Memory leaks may exist in the processes that were
+previously masked by doing a full shutdown and restart loop. 
+
+- Severity: Medium high
+- Controllers will continue to function (potentially in degraded state due to
+  lack of resources), and may be restarted frequently. However, cluster should
+  continue to function.
+
+Risk 2: Wedged KCM: There is a risk that controllers and the scheduler are not
+properly respecting context shutdowns. This can either result in multiple
+instances of controllers running or no instances running despite the lock being
+held.
+
+- Severity: High
+- Breaking mutual exclusion guarantees can put the cluster into a non-desirable
+  state. A manual user intervention is possible but if the problem is triggered
+  due to a problematic component, the issue will resurface and the best path for
+  mitigation is to turn off the feature.
+
+Risk 3: Futureproofing: An additional risk is that even if all the current code
+is safe and respects shutting down gracefully, new controllers/modifications to
+kcm or scheduler could create subtle problems in shutdown and transition.
+
+- Severity: High
+- Leads to either risk 1 or 2.
+
+
+Mitigations:
+
+- Audit and add tests for the existing controllers and the scheduler to ensure
+  proper handling of context shutdowns. See test plan section for more details.
+- Graceful shutdown modifications will not be guarded by a feature gate, but the
+  code change to remove the `os.Exit()` line will be guarded by a feature gate.
+- Document the new development best practices for graceful shutdown requirements
+  for modified components that are leader elected.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The core of this change involves modifying the `OnStoppedLeading` callback to
+not forcefully exit.
```

Comment:
I remember past issues with contexts not being well propagated to every worker of every controller.  Pod scaling in particular used separate workers and in the past they weren't directly waited upon.

Reply:
We can verify this with a VAP right? See if there are writes even after a leader loses its lock.